### PR TITLE
add editor to input options

### PR DESCRIPTION
**Pull request by Manyakalra3**
## Summary
- Added quilljsEditor to input options in `itemEditor` to enable formatting features in options too.

## TODOs

- [x] set border-width of all textarea.
- [x] set placeholder position correctly.
- [x] scrollbar should only appear when content exceeds size of tetarea.
- [x] remove extra space in textareas.
- [x] formatting icons should work properly with keys

## Test Plan

- [x] Test Responsiveness
   - [x] Laptop (1200px)
   - [x] Tablet (760px)
   - [x] Phone (320px)
- [ ] Cross-Browser Testing
   - [x] Chrome
   - [ ] Firefox
   - [ ] Safari (use BrowserStack)
- [ ] ~Local Language Support~
- [ ] Hygiene and Housekeeping
   - [x] Self-review
   - [x] Comments have been added appropriately
   - [x] Check for bundle size [here](https://bundlephobia.com/) if adding a package
   - [ ] Added relevant details like Labels/Projects/Milestones etc.
   - [ ] ~If adding or removing any environment variable:~
       - [ ] update `docs/ENV.md`
       - [ ] update Github Workflow files
       - [ ] update DockerFile
       - [ ] update the secrets for staging and production
- [ ] Testing
   - [ ] Wrote tests
   - [x] Tested locally
   - [ ] Tested on staging
   - [ ] Tested on an actual physical phone
   - [ ] Tested on production
- [ ] ~Lighthouse Checks~
   - [ ] Images have `alt` attributes
   - [ ] Any `` tags have `width` and `height` specified
   - [ ] Any `target="_blank"` links have `rel="noopener"`
   - [ ] Only SVGs are used as images. If PNGs are used, their size has been optimised.
   - [ ] Any SVGs without text have their `aria-label` attributes set

### CODE REVIEW

**Comment by deepansh96** [comment]
- In the question textbox and option textboxes, when we have formatting enabled, the formatting toolbar's top right and left edges are not rounded and that's why when selecting the textbox, you can see that you get this weird corner here.

https://user-images.githubusercontent.com/20169829/155678490-d7ae9a23-e759-46d1-820d-2d497ad1ed3b.mov

<hr>

- The placeholder's positioning is not correct in textboxes. In the question textbox, the placeholder is way towards the right, and the cursor starts from the very left. In the case of option textboxes, the placeholder is left of where the cursor starts

https://user-images.githubusercontent.com/20169829/155678851-6e0afb0d-3494-4a1a-8c48-61d5df6d7429.mov

<hr>

- By default, when there's nothing written in the question textbox, there is currently a small scrollbar. Which is not ideal because it gives the illusion that there is something to see whereas there is nothing. So a scrollbar should only appear when there is actually some content that exceeds the area of the textbox.

Screenshot 2022-02-25 at 1 44 17 PM

<hr>

- Why this extra space is there at the bottom of all the textboxes?
image

<hr>

- I think there's some fundamental bug with the implementation. Because when I'm typing, if I press Ctrl + B (to activate bold mode), and I start typing, something weird happens, you can see in the video below. This thing happens only if I press Ctrl + B/I/U and start typing. This DOES NOT happen for any other case, if I click the Bold/Italics/Underline icons, everything works flawlessly. Didn't you test it properly? You should test all the edge cases. If you don't, I'm anyway going to find it :P

https://user-images.githubusercontent.com/20169829/155693875-0b720924-dc49-4b4c-97f5-e6715cc488a8.mov

<hr>

- In the below video, I followed these steps. Clicked into the textbox -> Pressed "ctrl + B" on my keyboard -> I expected to see the Bold icon light up but it didn't. Same with the italics and underline. It's not that the bold mode or italics mode is not getting activated. It gets activated, but it's not visible on the icon. If I type any key on my keyboard after pressing "ctrl + B", it gets bold. Ideal behaviour should be -> Pressed Ctrl+B -> Bold icon gets lighted up.

https://user-images.githubusercontent.com/20169829/155695410-55ef4029-7c51-4bdf-ace0-4c8fc6c0069f.mov

<hr>

File: src/components/UI/Text/InputText.vue [hunk 1]
```diff
@@ -297,6 +345,48 @@ export default {
...
+        modules: {
+          toolbar: [["bold", "italic", "underline"]],
+        },
+        theme: "snow", //css for quilleditor
+        placeholder: this.placeholder, //placeholder for editor
+        formats: ["bold", "underline", "italic"], //formatting options for editor
+      });
+      this.quillEditor.root.innerHTML = this.value;
+      //invoked on input change
+      this.quillEditor.on("text-change", () => this.updateChange());
```
Comment (deepansh96) [inline src/components/UI/Text/InputText.vue:365]: The name of this method can be improved. 
The name of the method should also reflect what it does. This method emits the current value of the textbox. So try naming it similar to that

File: src/components/UI/Text/InputText.vue [hunk 2]
```diff
@@ -297,6 +345,48 @@ export default {
...
+}
+.ql-editor.ql-blank::before {
+  color: gray;
+  font-size: 0.9rem;
+  font-style: normal;
+}
+.textinput .ql-editor p {
+  margin-left: 12%;
+}
+</style>
```
Comment (deepansh96) [inline src/components/UI/Text/InputText.vue:392]: *to discuss*

File: src/components/UI/Text/Textarea.vue [hunk 3]
```diff
@@ -38,9 +41,7 @@
       </div>
       <!-- input text area -->
       <div
-        class="pt-7 overflow-y-auto h-32 border placeholder-blueGray-300 text-blueGray-600 bg-white rounded text-md border-blueGray-300 focus:outline-none focus:ring focus:border-transparent focus:shadow-outline w-full border-gray-200"
-        :class="[inputAreaClass, boxStyling]"
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:42]: You removed the `inputAreaClass` and `boxStyling` class?

File: src/components/UI/Text/Textarea.vue [hunk 4]
```diff
@@ -206,15 +208,14 @@ export default {
   methods: {
     //update HTML on input change in textarea
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:209]: same comments as in `InputText.vue` component.
- write a proper docstring, using syntax `/** ... */`
- write what this method is exactly doing

File: src/components/UI/Text/Textarea.vue [hunk 5]
```diff
@@ -206,15 +208,14 @@ export default {
...
     updateChange() {
-      //invoked on input change
+      //invoked on input changeconsole.
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:211]: Be a little more descriptive
Write proper sentences
`this method is invoked the user types in the input box`

Also, re-verify for typos before sending the PR for review! There's a small typo here

File: src/components/UI/Text/Textarea.vue [hunk 6]
```diff
@@ -226,10 +227,13 @@ export default {
 
     keyPress(event) {
       // invoked by pressing a key
+
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:230]:
```suggestion
```

File: src/components/UI/Text/Textarea.vue [hunk 7]
```diff
@@ -246,6 +250,7 @@ export default {
         toolbar: [["bold", "italic", "underline"]],
       },
       theme: "snow", //css for quilleditor
+      placeholder: this.placeholder, //placeholder for textarea
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:253]: no need of comment here

File: src/pages/Editor.vue [hunk 8]
```diff
@@ -2205,6 +2206,7 @@ export default {
         }
       }
       this.optionIndexToDelete = -1; // reset the option index to be deleted
+      this.reRender(); //for rerendering itemEditor whenever option is deleted.
```
Comment (deepansh96) [inline src/pages/Editor.vue:2209]:
```suggestion
      this.reRender(); // rerendering itemEditor whenever option is deleted.
```

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: src/components/UI/Text/InputText.vue [hunk 9]
```diff
@@ -25,11 +25,14 @@
       </div>
     </div>
 
-    <div class="flex relative mt-1 items-center">
+    <div
+      class="rounded textbox flex relative mt-1 items-center"
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue]: will move it to textarea.vue.

File: src/components/UI/Text/InputText.vue [hunk 10]
```diff
@@ -39,25 +42,42 @@
...
-        :disabled="isDisabled"
-        autocomplete="off"
-        data-test="input"
-      />
-
+      <div class="rounded w-full" v-if="isFormattingEnabled">
+        <div
+          class="textinput border placeholder-blueGray-300 text-blueGray-600 bg-white rounded text-md border-blueGray-300 focus:outline-none focus:ring focus:border-transparent focus:ring-primary focus:shadow-outline w-full overflow-ellipsis border-gray-200"
+          type="text"
+          name="placeholder"
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue]: This was actually part of the original code, we are passing props here,since it is a div now ,i will remove the unnecessary code.

File: src/components/UI/Text/InputText.vue [hunk 11]
```diff
@@ -70,6 +90,10 @@
 </template>
 
 <script>
+// importing css file is necessary and icons wont appear properly without it
+import Quill from "quill";
+import "quill/dist/quill.snow.css";
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:90]: okayyy

File: src/components/UI/Text/InputText.vue [hunk 12]
```diff
@@ -142,6 +171,11 @@ export default {
       default: false,
       type: Boolean,
     },
+    data() {
+      return {
+        editor: null,
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:171]: I will add comment here

File: src/components/UI/Text/InputText.vue [hunk 13]
```diff
@@ -269,6 +303,20 @@ export default {
     },
   },
   methods: {
+    // invoked on change in input
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:314]: Got it ,will make the changes

File: src/components/UI/Text/InputText.vue [hunk 14]
```diff
@@ -269,6 +303,20 @@ export default {
...
+    updateChange() {
+      this.$emit(
+        "update:value",
+        this.quillEditor.getText() ? this.quillEditor.root.innerHTML : "" //return the formatted value
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:318]: Yes,you asked this before,It is here just for handling any extra edge cases .should i remove it?

File: src/components/UI/Text/InputText.vue [hunk 15]
```diff
@@ -269,6 +303,20 @@ export default {
...
+      );
+    },
+    keyDown(event) {
+      // invoked by pressing a key
+      if (event.key == "Backspace" && this.quillEditor.root.innerText === "\n") {
+        event.preventDefault();
+      }
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:325]: done

File: src/components/UI/Text/InputText.vue [hunk 16]
```diff
@@ -297,6 +345,48 @@ export default {
     "update:value",
     "start-icon-selected",
     "end-icon-selected",
+    "keydown",
   ],
+
+  mounted() {
+    //new instance of quilljs is created
```
[the remark this answers is not in the document]
Comment (Manyakalra3) [reply src/components/UI/Text/InputText.vue:360]: done

File: src/components/UI/Text/InputText.vue [hunk 17]
```diff
@@ -25,11 +25,11 @@
       </div>
     </div>
 
-    <div class="flex relative mt-1 items-center">
+    <div :class="containerStyleClass">
```
Comment (deepansh96) [inline src/components/UI/Text/InputText.vue:28]: Refactor this file according to the changes made in Textarea.vue

File: src/components/UI/Text/Textarea.vue [hunk 18]
```diff
@@ -305,10 +305,10 @@ export default {
     },
     setToMaxHeight() {
       if (this.quillEditor != null)
-        this.quillEditorElement.style.height = `${this.maxHeightLimit}px`;
+        this.quillEditorElement.style.height = `${this.maxFHeightLimit}px`;
```
Comment (deepansh96) [inline src/components/UI/Text/Textarea.vue:308]: You didn't test this? Because it should've given an error

**Review by deepansh96 — CHANGES_REQUESTED** [review]
